third_party/nginx_latency_status_module/src/ngx_http_latency_recorder.c

```diff
@@ -0,0 +1,127 @@
+//    notice, this list of conditions and the following disclaimer.
+// 2. Redistributions in binary form must reproduce the above copyright
+//    notice, this list of conditions and the following disclaimer in the
+//    documentation and/or other materials provided with the distribution.
+//
+// THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
+// ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+// OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+// HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+// OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+// SUCH DAMAGE.
+////////////////////////////////////////////////////////////////////////////////
+
+#include "ngx_http_latency_recorder.h"
+#include "ngx_http_latency_storage.h"
+#include "ngx_http_latency_stub_status_module.h"
+
+
+// Get the index of the distribution bucket that a latency fall in.
+ngx_int_t get_latency_index(ngx_int_t max_exponent, ngx_int_t *latency_bucket_bounds, ngx_int_t latency) {
+  for (ngx_int_t i = 0; i <= max_exponent; i++) {
+    if (latency < latency_bucket_bounds[i]) {
+      return i;
+    }
+  }
+
+  // return the index for the overflow bucket.
+  return max_exponent + 1;
+}
+
+
+// Update a latency record with a single new latency.
+void update_latency_record(latency_stat *record, ngx_int_t latency, ngx_http_latency_main_conf_t *main_conf) {
+  ngx_int_t distribution_index;
+
+  ngx_atomic_fetch_add(&(record->request_count), 1);
+  ngx_atomic_fetch_add(&(record->sum), latency);
+  ngx_atomic_fetch_add(&(record->sum_squares), latency * latency);
+
+  distribution_index = get_latency_index(main_conf->max_exponent, main_conf->latency_bucket_bounds, latency);
+  ngx_atomic_fetch_add(&(record->distribution[distribution_index]), 1);
+}
+
+
+// Record the latency statistics.
+ngx_int_t ngx_http_record_latency(ngx_http_request_t *r)
+{
+  ngx_time_t *now;
+  ngx_msec_int_t request_time;
+  ngx_http_latency_conf_t *conf;
+  ngx_http_latency_main_conf_t *main_conf;
+  ngx_http_latency_shm_t *latency_record;
+
+  conf = ngx_http_get_module_loc_conf(r, ngx_http_latency_stub_status_module);
+
+  if (!conf->receiver_enabled) {
+    return NGX_DECLINED;
+  }
+
+  main_conf = ngx_http_get_module_main_conf(r, ngx_http_latency_stub_status_module);
+
+  // Update the cached current time value.
+  // This increases the accuracy of the latency measurment, but may slow down
+  // nginx. It may be removed after load testing.
+  ngx_time_update();
```

Comment:
This is my concern too.  It is a system call per request, it will slow down a lot.

Reply:
I was considering adding a flag to control whether this is called or not, so it can easily be turned on or off during testing. Do you think this is a good idea or are you suggesting that it should be removed now?

Reply:
Can we just remove it?  Do you really worry about the accuracy?   Nginx update cache time in every select() cycle.  It should not be too long.  Is that not enough?

Reply:
I've removed it.

I initially put it in because I kept getting latency measurements of 0, but this was probably because I was using either a hello_world module or the status page itself to test the latency measurements and they were genuinely taking < 1ms to run. I just tried adding a busy work for-loop to inflate the latency of the hello world module, and for that I get similar measurements with or without the update call.